### PR TITLE
Getting started with data preprocessing in Python

### LINKED ISSUE #2172: Getting started with data preprocessing in python
We ask that you please be patient as our team works through approving and publishing all articles/tutorials in a timely manner. Allow 1-3 days for a topic to be reviewed and/or approved - allow 3-7 days for an articles to be reviewed and/or published.

Be sure to visit our [Resources Page](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/resources-page.md) for tools, resources, and example articles to go over.

**NOTE: (Please ensure that you have only one open issue + linked pull request at a time. This will ensure that we complete the article in a timely manner from inception to publishing.)**

**NOTE: We tend to stray away or tend not to publish reviews/comparisons of commercial product offerings.**

*We're looking for the first 2-3 paragraphs of the article that appropriately summarize what your article will be about.*

### Proposed title of article
Getting started with data preprocessing in python

### Introduction paragraph (2-3 paragraphs):
Data preprocessing is one of the core steps in machine learning in which data is transformed into an appropriate format that makes it suitable to implement our models.
Data preprocessing involves 6 main steps.
These steps include importing the required libraries, importing the dataset, imputing the missing values, encoding categorical variables, splitting data into the training and testing sets, and finally features scaling.
Imputing missing values: this involves eliminating NA values from the data.
Encoding categorical variables: Involves transforming categorical variables into numeric values.
Splitting data set into training and test set: the training set is the portion of the data that we use to implement our model while the test set is the part we use to evaluate our model no
Feature Scaling: this involves transforming the units of the measurements of the features into standard units.
Feature are simply independent variables.

In this article, I'll explain what these steps are and also how to implement them in python. This will enable us to have appropriate data for future machine learning models.
### Key takeaways: 
1. Reader will know and appreciate the importance of data preprocessing as I'll explain everything as to why it's needed.
2. Reader will able to carry out data preprocessing activities in python.
3. Reader will have a clear understanding of the general activity.

### References: 
N/A

### Templates to use as guides
- [How To Guide Template](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/how-to-guide-template.md)
- [Software Review Template](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/software-review-template.md)
- [Tutorial Template](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/tutorial-template.md)

### PR DESCRIPTION
**Pull request by Madrinmarison**
### Article checklist

**ATTENTION: In your PR - add the ISSUE (topic form) # that this PR corresponds to. You can do this by also adding a comment that states **["This fixes #(enter your Idea Content Form #)"](https://youtu.be/TKJ4RdhyB5Y?t=278)** to link and close your Idea Content (issue) when the PR is merged.**
      
      
**NOTE: (Please ensure that you have only one open issue + linked pull request at a time. This will ensure that we complete the article in a timely manner from inception to publishing.)**

**If you have not already please go over our [Resources Page](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/resources-page.md) for more tips and tricks on how to improve our overall technical writing.**

#### Preliminary Checks - Formatting and Structure 
- [x] Does your article follow any of the suggested structure formats? - **see [example formats](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md#example-format-structure)**
- [x] Is your article properly formatted in Markdown? - **see [Github Markdown guide](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet)**
- [x] Have you used the correct folder and file structure? - **see [Contribution Guidelines](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/CONTRIBUTING.md)**
- [x] Is your article over 750 words? - **Tool**: [Wordcounter.net](https://wordcounter.net)
- [x] Have you used our preferred heading structure? - **[H3 (###) and up](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#headers)**
- [x] Does your article provide enough value and detail about your topic? - **Articles should be clear, accurate, and fully explained.**
- [x] Can your article be understood by beginners? - **Assume the audience is smart but has no prior exposure to the common terminology in your article.**
- [x] Have you included a hero image that is 600x400 pixels and have the copyright to use it? - **Tools:** [Pixlr Image Editor](https://pixlr.com/e) for resizing and [Unsplash](https://unsplash.com) for Creative Commons images
- [x] In your PR - add a comment that states **["This fixes #(enter your Idea Content Form #)"](https://youtu.be/TKJ4RdhyB5Y?t=278)** to link and close your Idea Content (issue) when the PR is merged.
      
#### Grammar & plagiarism checks
- [x] Have you spellchecked and grammar checked your article? - **Tools:** [Write&Improve](https://writeandimprove.com/?lang=en-GB) - [ProWritingAid](https://prowritingaid.com/) -  [Grammarly](https://grammarly.com) 
- [x] Please place your article through a 3rd party plagiarism checker? We suggest using Quetext, [this tool](https://www.quetext.com/) is free and has a daily limit. **We typically accept articles with 10% or less.**
- [x] Have you checked your article for readability? - **Tool:** [Hemmingway](http://www.hemingwayapp.com)
- [x] Have you added sources for quotes and images that aren't yours?

#### Technical checks
- [x] Are your code snippets properly formatted for syntax highlighting - **see [Syntax guide](https://gohugo.io/content-management/syntax-highlighting)**

### Contribution guidelines
For first-time contributors and for more details, **see [Contributing Guidelines](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/CONTRIBUTING.md)**

This fixes  #2172

### CODE REVIEW

**Comment by lalith1403** [comment]
Upon a first glance, this article requires a lot of changes grammatically. I suggest you look through the published articles to get a feel for the quality of the article we expect from the authors. 

Google is a proper noun and should be capitalized. google would be erroneous. The product's name is Google Colab, not colaboratory. The use of voice should be consistent throughout the article. 

Let me know when you rectify the entire article. Excited to review this one. :)

**Comment by Madrinmarison** [comment]
@lalith1403 I have made the requested changes.

File: content/articles/data-preprocessing-python/index.md [hunk 1]
```diff
@@ -0,0 +1,248 @@
...
+2. Connected to the internet.
+2. Access to Google Colab [Here](https://colab.research.google.com/notebooks/intro.ipynb).
+
+### Mounting our drive to Google Colab
+
+In this article, we shall carry out our data preprocessing activities on Google Colad. Therefore, we need to ensure our Google Drive is accessible by the Google Colab. To ensure this, first, let's download our data to our computer from [here](https://github.com/Madrinmarison/data).
+
+Since we've successfully downloaded our data, let's now upload it to google drive through [google.drive.com](https://drive.google.com/drive/u/0/my-drive).
+
+Now our data is uploaded to Google Drive.
```
Comment (lalith1403) [inline content/articles/data-preprocessing-python/index.md:34]: Explain the details of the folder in which you want the files saved for a later use.

File: content/articles/data-preprocessing-python/index.md [hunk 2]
```diff
@@ -0,0 +1,248 @@
...
+
+Above, we specified two variables, x for the features and y for the dependent variable. The features set, as declared in the code `Dataset.iloc[:, :-1]` will consist of all rows and columns of our dataset except the last column. Similarly, the dependent variable y will consist of all rows but only the last column as declared in the code `Dataset.iloc[:, -1].values`
+
+Let's have a look at our data by executing the code.
+
+```Python
+print(x) # returns an array of features
+```
+
+![Features_Array](/engineering-education/data-preprocessing-python/Feature.array.png)
```
Comment (lalith1403) [inline content/articles/data-preprocessing-python/index.md:86]: The terminal output can be shown as a formatted block instead of the screenshot.

For example, 

```bash
> x
[[1,2],[3,4]]
```

Just copy the terminal output and paste it in the formatted block.

File: content/articles/data-preprocessing-python/index.md [hunk 3]
```diff
@@ -0,0 +1,248 @@
...
+```Python
+print(y) # viewing an array of the dependent variable.
+```
+![Features_Array](/engineering-education/data-preprocessing-python/dependent_variable.png)
+
+Our data is imported successfully in the form of an array of features x and a dependent variable y.
+
+### Step 3: Taking care of the missing data
+
+We notice that x, an array of our features, contains missing values in the Age and Salary column from our output. We need to deal with this problem before we can implement our machine learning models on our data. There exist two techniques that we use to deal with the missing value:
```
Comment (lalith1403) [inline content/articles/data-preprocessing-python/index.md:97]: `Ages` and `Salary`. Use `` to highlight variables, column names etc.

File: content/articles/data-preprocessing-python/index.md [hunk 4]
```diff
@@ -0,0 +1,248 @@
...
+
+1. **Deleting the observation reporting the missing value(s):**
```
Comment (lalith1403) [inline content/articles/data-preprocessing-python/index.md:99]: Check grammar. Not sure what you want to convey.

File: content/articles/data-preprocessing-python/index.md [hunk 5]
```diff
@@ -0,0 +1,248 @@
...
+
+This technique works well with big datasets with few missing values. For instance, deleting a row from a dataset with hundreds of observations cannot affect the information quality of the dataset. However, this technique is not suitable for the dataset reporting many missing values. Deleting many rows from a dataset can lead to information loss information.
+
+To ensure no crucial information is lost, we need to make use of an appropriate technique.
+
+2. **Replace the missing value(s) with the average of the variable in which the data is missing:**
+
+This technique has reported no harm to the dataset's quality; therefore, it's the technique that we most use to handle missing values.
```
Comment (lalith1403) [inline content/articles/data-preprocessing-python/index.md:107]: How are you basing this statement? If you claim that filling missing value with averages does no harm to the dataset, provide justification. This is an important statement.

File: content/articles/data-preprocessing-python/index.md [hunk 6]
```diff
@@ -0,0 +1,248 @@
...
+
+Now that we know the techniques we use to take care of the missing values let's handle this problem in our dataset. Since our dataset is small, we cannot eliminate a row reporting the missing value(s). Therefore we shall make use of the second technique.
```
Comment (lalith1403) [inline content/articles/data-preprocessing-python/index.md:109]: We have only discussed one technique to replace the hidden values. There are many more available. You should discuss those as well.

File: content/articles/data-preprocessing-python/index.md [hunk 7]
```diff
@@ -0,0 +1,248 @@
...
+
+The `Country` and the Purchased columns of our dataset contain data that fall into categories. Since machine learning models are based on the mathematical equation, which takes only numerical inputs, it will be difficult for the machine learning model to compute the correlation between the feature and the dependent variables. To ensure this does not happen, we need to convert those string entries into numbers, i.e., we shall encode France into 0, Spain into 1, and Germany into 2. However, our future machine learning model can interpret the numerical order between 0 for France, 1 for Spain, and 2 for Germany do matter, which is not the case. To ensure this misinterpretation does not occur, we make use of one-hot encoding. One-hot encoding converts ours country column into three columns and creates a binary vector for each country such that there is no numerical order between the country categories. Therefore, instead of encoding the country variable into 1,2, and 3, vectors [1, 0 ,0] for the France, [0, 1, 0] for the Spain and [0, 0, 1] for the Germany. Similarly, we shall encode the Purchased column into 1,s for Yes and 0,s for No.
```
Comment (lalith1403) [inline content/articles/data-preprocessing-python/index.md:137]: Break this para into smaller paras.

File: content/articles/data-preprocessing-python/index.md [hunk 8]
```diff
@@ -0,0 +1,248 @@
...
+```python
+[0 1 0 0 1 1 0 1 0 1]
+
+```
+Our dependent variable is encoded successfully into 0's and 1's.
+
+### Step 5: Splitting the dataset into the training and test set
+
+In machine learning, we split the dataset into a training set and test set. The training set is the fraction of a dataset that we use to implement the model. On the other hand,  the test set is the fraction of the dataset that we use to evaluate the performance our the model. The test set is assumed to be unknown during the process of the model implementation.
+We need to split our dataset into four subsets, i.e., x_train, x_test, y_train, and y_test.
```
Comment (lalith1403) [inline content/articles/data-preprocessing-python/index.md:179]: `x_train`, `x_test`… Use these throughout the article, whenever a variable is referenced.

File: content/articles/data-preprocessing-python/index.md [hunk 9]
```diff
@@ -0,0 +1,248 @@
...
+```Python
+print(x_train)
+```
+![Scaled Feature](/engineering-education/data-preprocessing-python/scaled.png)
+
+```Python
+print(x_test)
+```
+
+```python
```
Comment (lalith1403) [inline content/articles/data-preprocessing-python/index.md:236]: This is a terminal output. Use bash instead of Python. 

``\`bash

**Review by lalith1403 — CHANGES_REQUESTED** [review]
@Madrinmarison Thanks for submitting an article on data preprocessing. When I read an article on data preprocessing, either I want it to cover breadth (tabular, images, audio, video, time series, etc) or I want it to go deep. 

Since you cover tabular data, I would expect you to add a lot more content in terms of the techniques used. When you talk about imputing values, talk about what happens to the distributions. I strongly suggest to check the claim that adding averages does not hurt.  

We usually also use data visualisation to check which parts of a dataset need to be processed and which can be left. Consider these points while building upon the article. Let this be the article that you will look to 6 months from now for data preprocessing of tabular data. 

_Additionally, split larger paragraphs into smaller ones._

**Comment by ninjaginja** [comment]
plagiarism check completed - all good
[2120420-9347b5204434e5a85081-1626789752149.pdf](https://github.com/section-engineering-education/engineering-education/files/6848877/2120420-9347b5204434e5a85081-1626789752149.pdf)

**Comment by lalith1403** [comment]
@Madrinmarison Have you made all the changes? I see only a few changes. Do let me know once done with all the changes.

Thank you.

**Comment by Madrinmarison** [comment]
@lalith1403  I've made all the changes.

**Comment by lalith1403** [comment]
@Madrinmarison Thank you for making the changes. However, the article is still not up to the mark for publishing. There are a few grammatical errors and it is time consuming for peer reviewers to rectify all the errors. 

I would suggest you to run the article through online services such as Grammarly. If possible, get it reviewed with a fellow EngEd contributor and update me. If the last option is not feasible, go through the published content for clarity on what we expect from articles. 

Look forward to getting this article published. 

Cheers,
Lalith

**Comment by Madrinmarison** [comment]
@lalith1403 I've made the requested changes.

**Comment by lalith1403** [comment]
@Madrinmarison There were many errors. Kindly go through the rectifications made. Incorporate them into your future articles to avoid article closures. Thank you.

**Comment by ahmadmardeni1** [comment]
I can't find a Hero image, can you upload it? @Madrinmarison

**Comment by ahmadmardeni1** [comment]
Thank you for submitting your article for publishing🚀 @Madrinmarison 
2 Important things:
1. I deleted the screenshots you provided, you can't give your token to anyone! They can access your data on your Drive account.
2. You didn't upload a Hero image, please upload it and I will publish your article.
Also, take a look at the changes I made.

Thank you for the review, please take a look at the changes I made for the front matter @lalith1403

**Comment by lalith1403** [comment]
> Thank you for submitting your article for publishing🚀 @Madrinmarison
> 2 Important things:
> 
> 1. I deleted the screenshots you provided, you can't give your token to anyone! They can access your data on your Drive account.
> 2. You didn't upload a Hero image, please upload it and I will publish your article.
>    Also, take a look at the changes I made.
> 
> Thank you for the review, please take a look at the changes I made for the front matter @lalith1403

@ahmadmardeni1 Thanks for finding the security issue with the token. I had thought the token was a unique one-time use-only phrase. I will make sure to ensure no keys are passed on. Thanks again.

**Comment by Madrinmarison** [comment]
@ahmadmardeni1 I have added hero image

**Comment by ahmadmardeni1** [comment]
Thank you and thank you for the review @lalith1403🚀

Great tutorial @Madrinmarison
